### ISSUE TITLE
Suspicious implementation of compare() - doesn't return 0 for equal objects 

### ISSUE THREAD

**Issue by dk2k** [quarkusio/quarkus#41733]
### Describe the bug

Line 18 in HasPriority
There is a reference to hashing mechanism in the comment: 
//they are functionally equal, but we don't want one to be discarded if there is a hash collision

But compare() is used for sorting in sorted collections. Not for hasing in hashed collections.
If the statement in comment is false, then maybe it's OK to return 0?

**Comment by FroMage** [comment]
I am frankly not sure what this thing does :-/

**Comment by dmlloyd** [comment]
This does seem like it could lead to an unstable sort (i.e. `a.compareTo(b) == 1` yet `b.compareTo(a) == 1` also).

I think the error here is in using a `Map` to order interceptors. It may be better to use e.g. `List#sort()` instead, which allows duplicates and is a stable sort.

**Comment by FroMage** [comment]
I don't doubt it's wrong. I'm not sure what it's used for, and how its wrongness manifests. A reproducer would help :)

**Comment by dk2k** [comment]
@FroMage That's a reproducer https://gist.github.com/dk2k/b69e54a38f58633879f579c85715bb66 . Don't forget to set a proper package name.
I wasn't sure what exactly I was supposed to reproduce. So I just created a Comparator with inconsistent logic - it doesn't return 0 for equal objects. The difference in the results can be seen if you change it to return 0 in the last line.
Currently the output is:
Element{intField=1}
Element{intField=0}
[Element{intField=2}]
The expected results are in comments.

**Comment by FroMage** [comment]
Oh, I meant a reproducer that shows how you can run into this bug using regular user code with Quarkus REST. I suppose you need to create two resource interceptors with the same priority, but again, I'm really unsure how the bug manifests and what it causes.

**Comment by geoand** [comment]
@dk2k any luck with what @FroMage asked for?

**Comment by dk2k** [comment]
@geoand No, I'm not going to do anything else in the scope of this issue. It either is mered as is, or is discarded.
